brianlong/www.validators.app
# 184383182 api limit rack attack

#### What's this PR do?
- Limit requests coming from API users

#### How should this be manually tested?
To test this locally, you need to run `rails dev:cache` to enable caching, and you need to have **redis** running. But I recommend to test this feature mostly on staging.
- Click through the app on staging, confirm that every page loads correctly
- Test the pages that uses websockets, eg. homepage, ping thing page, stake pools page. Confirm that they're working correctly and data is refreshed where should be refreshed.
- Send some POST requests to Ping Thing in small time intervals (more than 20). Confirm that you're not blocked. Confirm that ping thing page in browser is updated live.
- Send some other API GET requests in small time intervals (eg. https://stage.validators.app/api-documentation?locale=en&network=mainnet#ping), compare the responses. Try to send enough of them to get blocked. Then test that GET request with higher limit are still available - so send some request eg. to get Ping Thing list (https://stage.validators.app/api-documentation?locale=en&network=mainnet#ping-thing-list). 
- Wait 5 minutes and send request to ping-pong again. Confirm that you're no longer blocked.

#### What are the relevant tickets?
- [URL to the PivotalTracker ticket](https://www.pivotaltracker.com/story/show/184383182)

#### Task completed checklist:
- [] Is there appropriate test coverage?
- [] Did I take into consideration possible security vulnerabilities?
- [] Are the controllers secure?
- [] Was every important matter documented?

#### Things to watch out for
- [N] Does this PR have any migrations?
- [N] Does this PR add new dependencies?
